text/0027-corejam.md

```diff
@@ -0,0 +1,706 @@
+
+_(Note: depending on future Relay-chain Coretime scheduling implementation concerns, a window of Relay-chain blocks)._
+
+The need of validators to be rewarded for doing work they might reasonably expect to be useful competes with that of the Coretime procurers to be certain to get work done which is useful to them. In Polkadot 1.0, validators only get rewarded for PoVs ("work packages") which do not panic or overrun. This ensures that validators are well-incentivized to ensure that their computation is useful for the assigned parachain. This incentive model works adequately where all PVF code is of high quality and collators are few and static.
+
+However with this proposal (and even the advent of on-demand parachains), validators have little ability to identify a high-quality Work Package builder and the permissionless design means a greater expectation of flawed code executing in-core. Because of this, we make a slightly modified approach: Work Packages must have a valid Authorization, i.e. the Coretime-assigned `is_authorized` returns `true` when provided with the Work Package. However, Validators get rewarded for *any* such authorized Work Package, even one which ultimately panics or overruns on its evaluation.
+
+This ensures that Validators do a strictly limited amount of work before knowing whether they will be rewarded and are able to discontinue and attempt other candidates earlier than would otherwise be the case. There is the possibility of wasting Coretime by processing Work Packages which result in error, but well-written authorization procedures can mitigate this risk by making a prior validation of the Work Items.
+
+### Refine
+
+The `refine` function is implemented as an entry-point inside a code blob which is stored on-chain and whose hash is associated with the Work Class.
+
+```rust
+type ClassCodeHash = StorageMap<ClassId, CodeHash>;
+```
+
+```rust
+type WorkOutputLen = ConstU32<1024>;
+type WorkOutput = BoundedVec<u8, WorkOutputLen>;
+fn refine(payload: WorkPayload) -> WorkOutput;
+```
+
+Both `refine` and `is_authorized` are only ever executed in-core. Within this environment, we need to ensure that we can interrupt computation not long after some well-specified limit and deterministically determine when an invocation of the VM exhausts this limit. Since the exact point at which interruption of computation need not be deterministic, it is expected to be executed by a streaming JIT transpiler with a means of approximate and overshooting interruption coupled with deterministic metering.
+
+Several host functions (largely in line with the host functions available to Parachain Validation Function code) are supplied. Two additional ones include:
+
+```rust
+/// Determine the preimage of `hash` utilizing the Relay-chain Storage pallet.
+fn lookup(hash: [u8; 32]) -> Vec<u8>;
```

Comment:
It's worth noting that the storage pallet only guarantees items will be available _going forward_ from a minimum point in time. Therefore, this can lead to consensus splits unless a minimum relay-chain block for execution is specified (presumably in the `Context`).

Reply:
I presume this is clear from the Basic Validation description?